examples/cis2-wccd-upgradable/src/lib.rs

```diff
@@ -0,0 +1,3181 @@
+//! As follows from the CIS2 specification, the contract has a `transfer`
+//! function for transferring an amount of a specific token type from one
+//! address to another address. An address can enable and disable one or more
+//! addresses as operators. An operator of some token owner address is allowed
+//! to transfer any tokens of the owner.
+//!
+//! Besides the contract functions required CIS2, this contract implements a
+//! function `wrap` for converting CCD into wCCD tokens. It accepts an amount of
+//! CCD and mints this amount of wCCD. The function takes a receiving address as
+//! the parameter and transfers the amount of tokens.
+//!
+//! The contract also implements a contract function `unwrap` for converting
+//! wCCD back into CCD. The function takes the amount of tokens to unwrap, the
+//! address owning these wCCD and a receiver for the CCD. If the sender is the
+//! owner or an operator of the owner, the wCCD are burned and the amount of
+//! CCD is sent to the receiver.
+//!
+//! The protocol consists of three smart contracts (`proxy`, `implementation`,
+//! and `state`). All wccd functions (e.g. `wrap`, `unwrap`, `transfer`,
+//! `updateOperator`, `balanceOf`, `operatorOf` and `tokenMetadata`) have to be
+//! invoked on the `proxy` contract. The `proxy` will set the `current_sender`
+//! on the `implementation` before invoking these functions. Invoking the wccd
+//! functions directly on the `implementation` without going through the
+//! `proxy` fallback function will revert.
+
+//! The admin address on the `proxy` can upgrade the protocol with
+//! a new `implementation` contract. The admin address on the
+//! `implementation` can pause/unpause the protocol.
+//!
+//! Deploy the `state` and the `implementation` contract first by invoking their
+//! respective `init` functions. Then, deploy the `proxy` contract and pass
+//! the already deployed contract addresses into the proxy `init` function.
+//! Then, call the `initialize` functions on the `state` as well as on the
+//! `implementation` contracts to set the remaining addresses.
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use concordium_cis2::*;
+use concordium_std::{fmt::Debug, *};
+
+/// The id of the wCCD token in this contract.
+const TOKEN_ID_WCCD: ContractTokenId = TokenIdUnit();
+
+/// The metadata url for the wCCD token.
+const TOKEN_METADATA_URL: &str = "https://some.example/token/wccd";
+
+// Types
+
+/// Contract token ID type.
+/// Since this contract will only ever contain this one token type, we use the
+/// smallest possible token ID.
+type ContractTokenId = TokenIdUnit;
+
+/// Contract token amount type.
+/// Since this contract is wrapping the CCD and the CCD can be represented as a
+/// u64, we can specialize the token amount to u64 reducing module size and cost
+/// of arithmetics.
+type ContractTokenAmount = TokenAmountU64;
+
+/// Needed for the custom serial instance, which doesn't include the `Option`
+/// tag and the length of the vector.
```

Comment:
I took this part of the code from the “proxy” example that I found in the repo. The return value from the invoke in the fallback function (“host.invoke_contract_raw().1”) is an "Option" and it is converted into the “RawReturnValue”. If I understand correctly then we need this conversion from “return_value” into “RawReturnValue” because the fallback function needs to specify what type it returns. “ReceiveResult<RawReturnValue>” is a newly defined data type that has an unlimited vector length so we can pass anything from “return_value” into “RawReturnValue”.
I am not exactly sure what the `Option` tag refers to in the comment so.

Reply:
The example proxy contract is built so it also works with v0 contracts. Since v0 contracts don't have a return value, the option is needed.
I expect that we will only use v1 contracts in this case, so we can remove the option.

Reply:
How can you specify that this invoke is to a `version1` contract `host.invoke_contract_raw()?.1`? Currently, the `return_value` is an option and that is why we need to deal with the case that `None` is returned by this invoke. The `RawReturnValue` deals with both cases that `None` is returned or that some `return_value` is returned.

Reply:
You can't specify it directly, but v1 contracts will _always_ return a value (although it might be `()`), so if you unwrap the return value, we can get rid of the option. (I've sent more details directly to you).